### PR TITLE
#5090 finalise requisition with zero quantity lines

### LINKED ISSUE #5090: Request requisition: Keep requisition lines with zero requested quantity on finalise
## Is your feature request related to a problem? Please describe.

Some stores in Ivory Coast need to create request requisition lines even if the requested quantity is zero so we can push item data values to eSigl.

We will add a store pref [Store pref](https://github.com/sussol/msupply/issues/12273): "Keep requisition lines with zero requested quantity on finalise" to allow the creation of requisition lines with 0 quantity after finalise

this is not applicable for emergency orders

## Describe the solution you'd like

**A clear and concise description of what you want to happen.**

## Implementation

**A description of how the solution would best be implemented in the code base**

## Describe alternatives you've considered

**A clear and concise description of any alternative solutions or features you've considered.**

## Additional context

Dead line for PR : 10th Friday 2023

## Job code
[NFM_ESIGL] CI NFM3 : ESIGL Integration

### PR DESCRIPTION
**Pull request by arjunSussol**
Fixes #5090 

## Change summary

Able to finalize requisition with zero quantity lines if store preference `Keep requisition lines with zero requested quantity on finalise` is enabled. Also, this should not be applied to emergency order type program

## Testing

Steps to reproduce or otherwise test the changes of this PR:

- [ ] Enable the  store preference `Keep requisition lines with zero requested quantity on finalise` in desktop
- [ ] In mobile, create a general requisition with many lines
- [ ] Enter `requested quantity` value to some of the lines and keep zero for remaining.
- [ ] Click finalize button, all lines should be kept in the requisition.
- [ ] Repeat same by disabling the above preference, zero lines will be deleted from requisition.
- [ ] Create an emergency order type program requisition by enabling above preference.
- [ ] It should delete all lines having zero requested quantity.
- [ ] Repeat same for ordinary program requisition by enabling preference, zero lines will be shown.
- [ ] Repeat same by disabling preference with emergency and ordinary program requisition, delete all zero lines

### Related areas to think about

If there are any general areas of the codebase your changes might have side affects on, mention them here

### CODE REVIEW

File: src/database/DataTypes/Requisition.js [hunk 1]
```diff
@@ -469,15 +481,31 @@ export class Requisition extends Realm.Object {
...
+
   /**
    * Finalise this requisition.
    *
    * @param  {Realm}  database
    */
   finalise(database) {
-    this.pruneRedundantItems(database);
-    this.status = 'finalised';
+    const requisitionWithZeroQuantity = UIDatabase.getPreference(
```
Comment (sworup) [inline src/database/DataTypes/Requisition.js:499]: Nit-picking but this is a pref and should say `isKeepRequisitionWithZeroQuantityActive` or something.

**Review by sworup — APPROVED** [review]
The code looks good, just nit-picking a few things

**Comment by arjunSussol** [comment]
Hi @JonnaQuismundo desktop PR https://github.com/sussol/msupply/pull/12290 in v5.07 is good as IC is in that version.
Also, IC using mobile v5.1.1 and we are not going to upgrade that so soon.
So, I have to rebase this PR to v5.1.1 and trying to do it.

I have created branch from tag v5.1.1, but unfortunately can't run mobile successfully with this version.
There are some issues as it is very older version, so I am going to fix those issues first to run mobile v5.1.1 for IC.
After that I will update this PR base on new branch `V5.1.1-IC` which will be kept for IC only.

**Comment by arjunSussol** [comment]
Hi @JonnaQuismundo once able to successfully run mobile v5.1.1 with IC data locally, I will try to upgrade it to latest version v8.6.2. If able to upgrade without issue or with minimal issue with quick fix then we can upgrade mobile for IC with v8.6.2.

Please, let me know if you have any suggestion over it.

**Comment by JonnaQuismundo** [comment]
@arjunSussol I honestly think we must use the latest mobile. I don't it's worth doing in v5.1.1. We can try the latest first.

There are only 6 store sites that might have that prefs on:
C1781 CSU-DM CONFESSIONNEL Soeur Cathérine de YOPOUGON
C1405 CMS CONFESSIONNEL Ange Gardien de ABOBO
C1780 CMS CONFESSIONNEL Mère Maria de COCODY (COCODY BI)
C1391 CMS CONFESIONNEL Ste Thérèse Enfant Jesus de KOUMA
C4028 INF Public Prison Civil D’Abengourou
C4080 CMS PUBLIC POLICE NATIONALE D’AB

**Comment by JonnaQuismundo** [comment]
> Also, requisition reasoning validation is failed in latest mobile which should be fixed before delivering to IC as they are using this feature.

@arjunSussol is there an existing issue for this?

**Comment by JonnaQuismundo** [comment]
@arjunSussol let's just use the latest build, v8.6.2.

**Comment by arjunSussol** [comment]
> > Also, requisition reasoning validation is failed in latest mobile which should be fixed before delivering to IC as they are using this feature.
> 
> @arjunSussol is there an existing issue for this?

@JonnaQuismundo issue created for this https://github.com/openmsupply/mobile/issues/5109

File: src/database/DataTypes/Requisition.js (hunk 1 above)
Comment (arjunSussol) [reply src/database/DataTypes/Requisition.js]: @sworup should be fixed by https://github.com/openmsupply/mobile/pull/5104/commits/f5662cae36b543ea6487e44ccbcfb51690aef840

**Pull request merged**